feat: accept other types of input

theodoreb opened the pull request:
Help reduce cruft in code using this library. It is now possible to use: 
```js
once('id', 'selector');
once('id', 'selector', context);
once('id', document.body);
once('id', document); // Will apply to document.documentElement
```

refs #2

theodoreb committed:
feat: accept other types of input

Help reduce cruft in code using this library.

refs #2

theodoreb committed:
feat: Allow the use of document

Treat document as an alias to document.documentElement.

refs #2

KayLeung commented on API.md:
```diff
@@ -9,7 +9,7 @@
 
 <dl>
-<dt><a href="#once">once(id, elements)</a> ⇒ <code>Array.&lt;Element&gt;</code></dt>
+<dt><a href="#once">once(id, input, [context])</a> ⇒ <code>Array.&lt;Element&gt;</code></dt>
```
any better naming than `input`? `target`? `elementQuery|queryElements` ?

theodoreb replied:
jQuery uses 'selector' for this type of thing.

KayLeung replied:
`selector` is good.

theodoreb committed:
chore: rename parameter and update documentation

Rename parameter to be more explicit.
Document the new ways of using the once input parameters.

refs #2

theodoreb commented:
@KayLeung when we pass a single element, do you think the once function should return a single element instead of an array? 

```js
// currently
once('hop', document.body) // Return: [document.body]
// Should we do? 
once('hop', document.body) // Return: document.body (not in an array)
```

theodoreb commented:
always returns an array, if necessary use destructuring or .shift() to get a single element.

Now about the document issue.
```js
once('id', document) // applies to document.documentElement but return document? 
```

theodoreb committed:
docs: explain dealing with a single element

Explain how to use destructuring or array methods to keep user code short.

refs #2

theodoreb commented:
Replying to myself. Always return an array and return the document.documentElement. Added docs and tests to explain why.

theodoreb committed:
chore: move check in if test

fabianx-ai commented:
@theodoreb +1 to always returning an array, because the new pattern you want to replace the chaining of .once() into a function is:

```js
var x = once(...);
// Early return if we already processed those elements.
if (x.len == 0) {
  return;
}
```

KayLeung commented:
> @KayLeung when we pass a single element, do you think the once function should return a single element instead of an array?
> 
> ```js
> // currently
> once('hop', document.body) // Return: [document.body]
> // Should we do? 
> once('hop', document.body) // Return: document.body (not in an array)
> ```

Consistently returns `ARRAY` is better. Simple API. I think some cases are non-predictable or debatable.

```JS
document.body // predictable, this is single

classSelector // non-predictable, how to handle single return?

IdButReturnedTwo // predictable but debatable because it still can return more than two.

[data-Drupal-style] // non-predictable one or more and debatable
```

theodoreb commented:
Last question: letting people use `document`.

I added th ability to use document to go around Drupal issue: https://www.drupal.org/project/drupal/issues/3160052 but I really don't like that it changes the element to work on.

theodoreb committed:
chore: remove support for document as parameter

Do not try and fix broken code, keep the library predictable

refs #2

theodoreb committed:
chore: remove support for document as parameter

Do not try and fix broken code, keep the library predictable

refs #2

theodoreb committed:
chore: optimize code for minification

theodoreb commented:
Removed the ability to use document, let's not start adding technical debt. fix on Drupal side is needed.

theodoreb committed:
docs: add required description

theodoreb committed:
chore: simplify call
